### PR TITLE
Inherit references from previous submissions

### LINKED ISSUE #6474: Reuse already loaded assemblies if they are identical
Scenario:
Solution with 3 projects that don't build into a common directory but rather each builds into their own:

```
Shared\bin\debug\Shared.dll
Lib1\bin\debug\Lib1.dll 
               Shared.dll (copy)
Lib2\bin\debug\Lib2.dll
               Shared.dll (copy)
```

Now in a REPL:

```C#
> #r "Lib1\bin\debug\Lib1.dll" 
> Lib1 l1 = new Lib1();  // loads Lib1 and Shared from Lib1\bin\debug dir
> #r "Lib2\bin\debug\Lib2.dll" 
> Lib2 l2 = new Lib2();   // loads Lib1 and Shared from Lib2\bin\debug dir
> l2.libBase
Field not found: 'Lib2.libBase'.
  + System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[TResult].Start(TStateMachine& stateMachine)
```

The problem is that we load Shared twice at runtime because it's in a different directory. The compiler correctly dedups it since both copies have the same identity. The loader should check whether the assemblies are the same. If they are then we can reuse the existing one. Otherwise we should report an error.

### PR DESCRIPTION
**Pull request by tmat**
Moves the logic of reusing metadata references from Script down to ReferenceManager. This is necessary to avoid duplication of PE assembly symbols due to repeatedly resolved missing references. 

The metadata resolver creates a new reference whenever asked to resolve a missing assembly identity. That results in creating new PE assembly symbol. We could implement deduplication in the resolver. However the lifetime of the resolver isn't necessarily related to the lifetime of the interactive session. hence it is better to implement it in the ReferenceManager. This work item was previously suggested for U1 but postponed for U2 (#5301) since the thinking was that it only affects performance. However, as described in #6457 it is critical for correctness of basic scenarios in csi and Hosting APIs, which don't use shadow copying.

Also fixes #6474.

### CODE REVIEW

**Comment by tmat** [comment]
@dotnet/interactive @dotnet/roslyn-compiler

**Comment by ManishJayaswal** [comment]
@tmat @genlu some testing ideas to try out if you haven't covered them already ( -> is depends on):
1. A -> B, B-> A, #r A
2. A->C, B->C, #r A, #r B
3. Same as 2 but two copies of C are in different folders and A and B refer to different copies
4. same as above by C is of different versions, C and C'
5. same as above where A, B, C and C' are in the same folder
6. circular A-> B -> C -> A, #r A
7. Missing assemblies are in gac as well as local folder
8. A -> C and C', #r A
9. A-> C and C', B-> C, #r B, #r A
10 A->C, #r A, #r C
11 A-> C, #r C, #r A

**Comment by ManishJayaswal** [comment]
:+1: pending buddy testing.

**Comment by tmat** [comment]
test this please -- i don't think this run thru CI as it was down.

**Comment by tmat** [comment]
test this please

**Comment by genlu** [comment]
It still has problem with the following scenario:
A -> C, B -> C, and each has its only copy of the same version of C in their own folders. "#r A" and "#r b" returns no errors but actually access the member of C in B throw a MissingFieldException.

In ther following example, `Lib1` and `Lib2` both reference to `LibBase`.

missingfield

File: src/Interactive/Features/Interactive/Core/InteractiveHost.Service.cs [hunk 1]
```diff
@@ -647,6 +650,9 @@ public void SetTestObjectFormattingOptions()
                                 {
                                     rspState = rspState.WithScriptState(newScriptState);
                                 }
+
+                                // remove references and imports from the options, they have been applied and will be inherited from now on:
+                                rspState = rspState.WithOptions(rspState.ScriptOptions.WithReferences(Array.Empty<MetadataReference>()).WithImports(Array.Empty<string>()));
```
Comment (cston) [inline src/Interactive/Features/Interactive/Core/InteractiveHost.Service.cs]: Perhaps add an ```EvaluationState.WithNoReferencesOrImports()``` helper method.

**Comment by cston** [comment]
LGTM

**Comment by tmat** [comment]
test win please

**Comment by tmat** [comment]
@genlu The issue is in the assembly loader. It doesn't dedup copy of LibBase from different directory.

**Comment by genlu** [comment]
@tmat @ManishJayaswal  https://github.com/dotnet/roslyn/issues/6474 is blocking us from more testing on this change. I think we should hold off the merge for this until https://github.com/dotnet/roslyn/issues/6474 is resolved.

**Comment by ManishJayaswal** [comment]
Yes I agree.

**Comment by tmat** [comment]
OK.

**Comment by tmat** [comment]
Added a commit that fixes https://github.com/dotnet/roslyn/issues/6474

**Comment by genlu** [comment]
@tmat Got another crash. The repro is the same as described above except `Lib1` and `Lib2` reference to different version of `LibBase`.

**Comment by tmat** [comment]
Crash? An error should be reported.

**Comment by genlu** [comment]
No, i didn't mean crash, just the same MissingFieldException. The same version of lib is fixed.

**Comment by tmat** [comment]
test this please

**Comment by tmat** [comment]
the 64bit failure is a time out. merging.

**Pull request merged**